File: web/src/api/consumerOffsetCopy.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0.
 */
import client from './client';

export interface CopySelection {
  instanceId: string;
  topic: string;
  sourceGroup: string;
  targetGroup: string;
}
export interface CopyExpected {
  brokerName: string;
  brokerAddr: string;
  queueId: number;
  sourceOffset: string;
  targetOffset: string | null;
}
export interface CopyPreview {
  topic: string;
  sourceGroup: string;
  targetGroup: string;
  queues: { expected: CopyExpected; minOffset: string; maxOffset: string }[];
}
export interface CopyReceipt {
  queues: {
    queue: CopyExpected;
    status: 'CONFIRMED' | 'UNCHANGED' | 'UNKNOWN' | 'NOT_ATTEMPTED';
    observedOffset: string | null;
  }[];
}
export async function previewConsumerOffsetCopy(params: CopySelection, signal?: AbortSignal) {
  const response = await client.get<{ data: CopyPreview }>('/consumer-offset-copy', {
    params,
    signal,
  });
  return response.data.data;
}
export async function applyConsumerOffsetCopy(
  selection: CopySelection,
  expected: CopyExpected[],
  signal?: AbortSignal,
) {
  const response = await client.post<{ data: CopyReceipt }>(
    '/consumer-offset-copy',
    { ...selection, expected },
    { signal, timeout: 0 },
  );
  return response.data.data;
}

File: server/src/main/java/org/apache/rocketmq/studio/instance/group/ConsumerOffsetCopy.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0.
 */
package org.apache.rocketmq.studio.instance.group;

import java.util.List;

public final class ConsumerOffsetCopy {
    private ConsumerOffsetCopy() { }

    public record Expected(String brokerName, String brokerAddr, int queueId,
            String sourceOffset, String targetOffset) { }

    public record Queue(Expected expected, String minOffset, String maxOffset) { }

    public record Preview(String topic, String sourceGroup, String targetGroup, List<Queue> queues) { }

    public record Request(String instanceId, String topic, String sourceGroup, String targetGroup,
            List<Expected> expected) { }

    public enum Status { CONFIRMED, UNCHANGED, UNKNOWN, NOT_ATTEMPTED }

    public record Outcome(Expected queue, Status status, String observedOffset) { }

    public record Receipt(List<Outcome> queues) { }
}
